### PR TITLE
Bug 1468943 - Mark confirmed alerts automatically

### PR DESCRIPTION
**Pull request by ionutgoldan**
@wlach I could use your assistance here. I'm having a perf alert + a list containing 5 pushes. The 3rd push is associated with the alert, while the other 4 are just neighboring pushes (2 before, 2 after the 3rd push).
This background worker I'm working on must go through all these 5 pushes and check to see if there are any pending jobs that must run/are running perf tests with the same signature as those that created the perf alert.
In short: how can I find the jobs that produced data points for a particular perf alert?

### CODE REVIEW

**Comment by wlach** [comment]
No easy way, we always look for a window (12-24 points before, 12 points after) for each data point in a series, and trigger an alert where the probability of a regression in an after window is high:

https://github.com/mozilla/treeherder/blob/6c24ff9f94f5a7b26b83797cbf690d65a5f57f1d/treeherder/perf/alerts.py#L36

This process leaves no specific trace of what data was considered. Probably your best bet for something like this is to look up adjacent pushes to an alert and then trigger talos/raptor/perfherder jobs for them.

File: treeherder/perf/tasks.py [hunk 1]
```diff
@@ -10,3 +13,61 @@ def generate_alerts(signature_id):
...
+        for push in pushes_to_check:
+            pending_perf_jobs_exist = Job.objects.filter(
+                push=push, job_type=job_type, autoclassify_status=Job.PENDING).exists()
+            if pending_perf_jobs_exist:
+                has_pending_jobs = True
+                break
+
+        if not has_pending_jobs:
+            alert.status = PerformanceAlert.CONFIRMED
+            alert.save()
```
Comment (wlach) [inline treeherder/perf/tasks.py]: You could simplify lines 63-73 to something like:

```py
if not Job.objects.filter(push__in=pushes_to_check, job_type=job_type, autoclassify_status=Job.PENDING).exists():
   alert.status = PerformanceAlert.CONFIRMED
   alert.save()
```

File: treeherder/perf/tasks.py [hunk 2]
```diff
@@ -10,3 +13,61 @@ def generate_alerts(signature_id):
     newrelic.agent.add_custom_parameter("signature_id", str(signature_id))
     signature = PerformanceSignature.objects.get(id=signature_id)
     generate_new_alerts_in_series(signature)
+
+
+@task(name='check-confirming-perf-alerts')
+def check_confirming_perf_alerts_status():
+    # TODO: what's this for? do I need it?
+    # newrelic.agent.add_custom_parameter("signature_id", str(signature_id))
```
Comment (wlach) [inline treeherder/perf/tasks.py]: This just annotates the new relic logs with extra data, I wouldn't worry about adding lines like this unless you're trying to debug a problem.

File: treeherder/perf/tasks.py [hunk 3]
```diff
@@ -10,3 +13,61 @@ def generate_alerts(signature_id):
...
+    one_day = timedelta(hours=24)
```
Comment (wlach) [inline treeherder/perf/tasks.py]: I would define this as something like `PERFHERDER_CONFIRMING_INTERVAL` in the settings.

File: treeherder/perf/tasks.py [hunk 4]
```diff
@@ -10,3 +13,61 @@ def generate_alerts(signature_id):
...
+            return []
+
+    confirming_alerts = PerformanceAlert.objects.filter(status=PerformanceAlert.CONFIRMING)
+    for alert in confirming_alerts:
+        alert_push = alert.summary.push
+        _from, _to = alert_push.time-one_day, alert_push.time+one_day
+
+        job_type = traceback_producer_job_type(alert)
+        if job_type is None:
+            # Warning
```
Comment (wlach) [inline treeherder/perf/tasks.py]: You should at least log a warning here, if not throw an exception (I would personally throw an exception)

**Comment by ionutgoldan** [comment]
@edmorley could you have a look over the Celery worker setup and see if it's ok?

File: treeherder/perf/tasks.py [hunk 5]
```diff
@@ -10,3 +16,49 @@ def generate_alerts(signature_id):
...
+
+        alert_push_idx = None
+        for index, push in enumerate(many_nearby_pushes):
+            if alert_push.id == push.id:
+                alert_push_idx = index
+                break
+
+        from_ = alert_push_idx - push_range
+        from_ = from_ if from_ > 0 else 0
+        to_ = alert_push_idx + push_range
```
Comment (wlach) [inline treeherder/perf/tasks.py]: Why the underscores after the variable names? I would personally probably name them `push_from` and `push_to`

File: treeherder/perf/tasks.py [hunk 6]
```diff
@@ -10,3 +16,49 @@ def generate_alerts(signature_id):
...
+                break
+
+        from_push = alert_push_idx - push_range
+        from_push = from_push if from_push > 0 else 0
+        to_push = alert_push_idx + push_range
+
+        return list(many_nearby_pushes[from_push:to_push])
+
+    confirming_alerts = PerformanceAlert.objects.filter(status=PerformanceAlert.CONFIRMING)
+    for alert in confirming_alerts:
```
Comment (edmorley) [inline treeherder/perf/tasks.py]: This looks like it could be expensive in terms of queries. Could we test on prototype first?

File: treeherder/perf/tasks.py [hunk 7]
```diff
@@ -10,3 +16,49 @@ def generate_alerts(signature_id):
...
+        job_type = traceback_producer_job_type(alert)
+        if job_type is None:
+            raise LookupError('Could not find job type for perf alert with id {}'.format(alert.id))
+
+        many_nearby_pushes = Push.objects.filter(
+            time__gte=_from, time__lte=_to, repository=alert_repository).order_by('time')
+        pushes_to_check = extract_pushes_to_check(many_nearby_pushes, alert_push)
+
+        if not Job.objects.filter(
+                push__in=pushes_to_check, job_type=job_type, autoclassify_status=Job.PENDING).exists():
```
Comment (edmorley) [inline treeherder/perf/tasks.py]: `autoclassify_status` means something else - and using it could give a race condition where an alert never gets confirmed. I think `state` is what you were after?

Also, what happens if a scheduled job doesn't run or gets stuck as pending forever (this happens from time to time). In that case it seems that the alert will get stuck as "confirming" and the number of alerts handled in this for loop will slowly increase over time, causing the cost of this task to increase.

Could the end goal of this feature instead be achieved in a simpler way, by instead adding a "waiting" status (which indicates the alert is not actionable yet, like "confirming" did), but then just have this task change "waiting" to "<something>" once a set amount of time has passed (eg 2 hours, which should be enough time for the jobs to have finished)?

File: treeherder/perf/migrations/0010_add_confirmed_status.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('perf', '0009_non_nullable_issue_tracker'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='performancealert',
```
[the remark this answers is not in the document]
Comment (ionutgoldan) [reply treeherder/perf/migrations/0010_add_confirmed_status.py]: >Do these result in actual MySQL schema migrations, or are they purely Django internal migrations? (ie: what does ./manage.py sqlmigrate ... say?)

Good question: I ran the command and no real migration actually happens. Guess I should just remove that migration script.

File: treeherder/perf/tasks.py [hunk 9]
```diff
@@ -10,3 +16,49 @@ def generate_alerts(signature_id):
     newrelic.agent.add_custom_parameter("signature_id", str(signature_id))
     signature = PerformanceSignature.objects.get(id=signature_id)
     generate_new_alerts_in_series(signature)
+
+
+@task(name='check-confirming-perf-alerts')
+def check_confirming_perf_alerts_status():
```
[the remark this answers is not in the document]
Comment (ionutgoldan) [reply treeherder/perf/tasks.py]: That sounds good.

File: treeherder/perf/tasks.py (hunk 6 above)
Comment (wlach) [reply treeherder/perf/tasks.py]: I don't think it should be very expensive, as should never have all that many confirming alerts afaik.

Comment (wlach) [reply treeherder/perf/tasks.py]: (though I agree testing on prototype is a good idea regardless)

File: treeherder/perf/tasks.py (hunk 7 above)
Comment (ionutgoldan) [reply treeherder/perf/tasks.py]: I like the background worker approach because it's more deterministic and you don't have to peak on every 'confirming' alert.

Comment (edmorley) [reply treeherder/perf/tasks.py]: The snooze feature would still be implemented as a background worker, if you thought I meant otherwise? :-)

File: treeherder/perf/alerts.py [hunk 10]
```diff
@@ -15,6 +17,51 @@
...
+
+        alert_push_idx = None
+        for index, push in enumerate(many_nearby_pushes):
+            if alert_push.id == push.id:
+                alert_push_idx = index
+                break
+
+        from_push = alert_push_idx - push_range
+        from_push = from_push if from_push > 0 else 0
+        to_push = alert_push_idx + push_range
```
Comment (jmaher) [inline treeherder/perf/alerts.py]: while we care about +-2 for a push range, we really care about +-2 existing data points- the difference is that push range can include pushes that have no data (i.e. needs a backfill.)  In this case we would need a larger range to consider.  Also in many cases we cannot backfill (broken builds, infra down, etc.) and this +-2 is going to fail and be confusing as our edge cases will pile up.

File: treeherder/perf/alerts.py [hunk 11]
```diff
@@ -15,6 +17,51 @@
...
 
+def check_confirming_perf_alerts_status():
+    confirming_interval = settings.PERFHERDER_CONFIRMING_INTERVAL
+
+    def traceback_producer_job_type(perf_alert):
+        push = perf_alert.summary.push
+        signature = perf_alert.series_signature
+
+        associated_perf_datum = PerformanceDatum.objects.filter(push=push, signature=signature).first()
+        return associated_perf_datum.job.job_type if associated_perf_datum else None
```
[the remark this answers is not in the document]
Comment (ionutgoldan) [reply treeherder/perf/alerts.py]: @jmaher I don't really understand what's to address here. Could you explain more?

[the remark this answers is not in the document]
Comment (jmaher) [reply treeherder/perf/alerts.py]: I was offering a suggestion for more hacking on this PR, or to open a followup PR

[the remark this answers is not in the document]
Comment (ionutgoldan) [reply treeherder/perf/alerts.py]: > Outside of those comments, I see another change I would like to see. We would now be adding Yet Another Confusing State (YACS) to the sheriffing workflow. I find it difficult already to toggle between untriaged and confirming for the long list of frameworks we support. Can we just assume everything in untriaged is confirming and only moved to confirmed when we have retrigger/backfill data?

Ok, sounds good. This means we must only rename the 'confirming' state to 'confirmed'.

[the remark this answers is not in the document]
Comment (ionutgoldan) [reply treeherder/perf/alerts.py]: There's a distinction between 'untriaged' and 'confirming'.

[the remark this answers is not in the document]
Comment (ionutgoldan) [reply treeherder/perf/alerts.py]: @jmaher We could turn the 'confirming'/'confirmed' states into simple labels per alert. These labels will work just like the star mechanism. This addresses 2 problems: the YACS + the worker not knowing what alerts to check for.
What do you think?

[the remark this answers is not in the document]
Comment (jmaher) [reply treeherder/perf/alerts.py]: interesting thought- I like the ideas of labels better than states- in fact if we can reduce a few states and use labels, that would be a win.  Given the fact that we will use dedicated perf sheriffs for our workflows, I would like to assume the dedicated sheriffs will be mindful of resources and choose the best alerts to work from- then we can focus on changes that support >1 sheriff instead of changes that also support sheriffs who are not trained as much.

**Comment by edmorley** [comment]
Closing this out for now since [bug 1468943](https://bugzilla.mozilla.org/show_bug.cgi?id=1468943) has been unassigned/deferred to the future.
